Properly serialize types that only appear at function input

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47775 Properly serialize types that only appear at function input**

When serializing graphs, we check every node for named types referenced,
so that we can register them as dependencies. We were skipping this
check for the graph inputs themselves. Since types used at input are
almost always used somewhere in the graph, we never noticed this gap
until a user reported an issue with NamedTuples.

Differential Revision: [D24896289](https://our.internmc.facebook.com/intern/diff/D24896289)

suo committed:
Properly serialize types that only appear at function input

When serializing graphs, we check every node for named types referenced,
so that we can register them as dependencies. We were skipping this
check for the graph inputs themselves. Since types used at input are
almost always used somewhere in the graph, we never noticed this gap
until a user reported an issue with NamedTuples.

[ghstack-poisoned]

jamesr66a reviewed (commented):
What about at the output? I can sneakily construct a counterexample like so:

```
import torch
from typing import NamedTuple, Optional

class Point(NamedTuple):
  x : int
  y : int

class Foo(torch.nn.Module):
  def __init__(self):
    super().__init__()

  def forward(self) -> Optional[Point]:
    return None

script = torch.jit.script(Foo())
torch.jit.save(script, 'foo.zip')
torch.jit.load('foo.zip')

"""
RuntimeError: 
Unknown type name '__torch__.Point':
Serialized   File "code/__torch__.py", line 5
  __buffers__ = []
  training : bool
  def forward(self: __torch__.Foo) -> Optional[__torch__.Point]:
                                               ~~~~~~~~~~~~~~~ <--- HERE
    return None


"""
```

suo committed:
Update on "Properly serialize types that only appear at function input"


When serializing graphs, we check every node for named types referenced,
so that we can register them as dependencies. We were skipping this
check for the graph inputs themselves. Since types used at input are
almost always used somewhere in the graph, we never noticed this gap
until a user reported an issue with NamedTuples.

Differential Revision: [D24896289](https://our.internmc.facebook.com/intern/diff/D24896289)

[ghstack-poisoned]

suo committed:
Update on "Properly serialize types that only appear at function input"


When serializing graphs, we check every node for named types referenced,
so that we can register them as dependencies. We were skipping this
check for the graph inputs themselves. Since types used at input are
almost always used somewhere in the graph, we never noticed this gap
until a user reported an issue with NamedTuples.

Differential Revision: [D24896289](https://our.internmc.facebook.com/intern/diff/D24896289)

[ghstack-poisoned]

jamesr66a reviewed (approved):
![ezgif com-gif-maker](https://user-images.githubusercontent.com/4685384/98863193-dff52000-241c-11eb-9c6b-5f6ce605117f.gif)
